crates/sui-archival/src/blob.rs

```diff
@@ -0,0 +1,549 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Context, Result};
+use byteorder::{BigEndian, ByteOrder, ReadBytesExt};
+use crc::{Crc, CRC_32_ISCSI};
+use integer_encoding::VarInt;
+use integer_encoding::*;
+use num_enum::IntoPrimitive;
+use num_enum::TryFromPrimitive;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs::{read_dir, File, OpenOptions};
+use std::io::{BufReader, BufWriter, Read, Seek, SeekFrom, Write};
+use std::path::PathBuf;
+use std::{fs, mem};
+use tracing::info;
+use tracing::log::error;
+
+/// The following describes the format of a blob file, which is created in the `transactions/`,
+/// `/effects` or `/objects` directory of an archive. The maximum size per file is 32MB. Blobs (
+/// transactions, effects or objects) in the files are referenced from the index by uint64 composed
+/// of in-file offset (lower 4 bytes) and file sequence number (upper 4 bytes). Blobs are written
+/// out to local disk and shipped to remote storage optionally. A checkpoint tailer will subscribe
+/// to and continuously archive transactions and effects as blobs. For state snapshot, objects will
+/// be archived at the end of every epoch by scanning the checkpointed objects db table. One
+/// checkpoint directory could have data for multiple checkpoints in case of transactions and effects.
+/// For objects, a snapshot would be taken at the end of epoch and stored under the end of epoch
+/// checkpoint.
+/// Sui Archive Directory Layout
+///  - archive/
+///     - transactions/
+///       - checkpoint_000001/
+///         - 000001.blob
+///         - 000002.blob
+///         - 000003.blob
+///         - ...
+///         - 020000.blob
+///         - index
+///       - checkpoint_000100/
+///         - 000001.blob
+///     - effects/
+///       - checkpoint_000001/
+///         - 000001.blob
+///         - 000002.blob
+///         - ...
+///         - index
+///       - checkpoint_000050/
+///         - 000001.blob
+///         - 000002.blob
+///         - ....
+///         - index
+///     - objects/
+///       - checkpoint_100000/
```

Comment:
I thought having it by checkpoints provide an opportunity to save snapshots at any random checkpoint (not just epoch ending one). But I am ok to move to epoch# for objects

Reply:
i suppose we can do that as well, clients will always have both the epoch and checkpoint sequence number anyway. However I don't think we will ever restore at non-epoch boundaries, so it may be simpler to stick with epochs.